src/SubscribedStreamPartition.js

```diff
@@ -1,92 +1,102 @@
+            getStream: this.getStream,
+            isPublisher: memoize(async (publisherId, _streamId) => (
+                this._client.isStreamPublisher(_streamId, publisherId)
+            ), memoizeOpts),
+            isSubscriber: memoize(async (ethAddress, _streamId) => (
+                this._client.isStreamSubscriber(_streamId, ethAddress)
+            ), memoizeOpts),
+        })
+        this.getPublishers = memoize(this.getPublishers.bind(this), memoizeOpts)
+        this.getSubscribers = memoize(this.getSubscribers.bind(this), memoizeOpts)
+        this.isValidPublisher = memoize(this.isValidPublisher.bind(this), memoizeOpts)
+        this.isValidSubscriber = memoize(this.isValidSubscriber.bind(this), memoizeOpts)
     }
 
     async getPublishers() {
-        if (!this.publishersPromise || (Date.now() - this.lastUpdated) > PUBLISHERS_EXPIRATION_TIME) {
-            this.publishersPromise = this._client.getStreamPublishers(this.streamId).then((publishers) => {
-                const map = {}
-                publishers.forEach((p) => {
-                    map[p] = true
-                })
-                return map
+        const publishers = await this._client.getStreamPublishers(this.streamId)
+        return publishers.reduce((obj, key) => (
+            Object.assign(obj, {
+                [key]: true
             })
-            this.lastUpdated = Date.now()
-        }
-        return this.publishersPromise
+        ), {})
     }
 
-    async _isPublisher(publisherId) {
-        if (!this.isPublisherPromises[publisherId]) {
-            this.isPublisherPromises[publisherId] = this._client.isStreamPublisher(this.streamId, publisherId)
-        }
-        return this.isPublisherPromises[publisherId]
+    async getSubscribers() {
+        const subscribers = await this._client.getStreamSubscribers(this.streamId)
+        return subscribers.reduce((obj, key) => (
+            Object.assign(obj, {
+                [key]: true
+            })
+        ), {})
     }
 
     async isValidPublisher(publisherId) {
-        const cache = await this.getPublishers()
-        if (cache[publisherId]) {
-            return cache[publisherId]
+        return this._client.isStreamPublisher(this.streamId, publisherId)
+    }
+
+    async isValidSubscriber(ethAddress) {
+        return this._client.isStreamSubscriber(this.streamId, ethAddress)
+    }
+
+    async _verifyStreamMessage(msg) {
+        try {
+            await this.validator.validate(msg)
+        } catch (err) {
+            // store error for possible introspection later
+            // doesn't ever clear value because can't be async safe without complication
+            this.lastValidationError = err
+            return false
         }
-        const isValid = await this._isPublisher(publisherId)
-        cache[publisherId] = isValid
-        return isValid
+        return true
     }
 
     async verifyStreamMessage(msg) {
```

Comment:
This method needs a bit of a rewrite:

- We should **always** validate key exchange-related messages, regardless of the `verifySignatures` option! So we need to inspect the type of the message and only apply the `verifySignatures` policy in case it's a normal message.
- The stream's policy (`stream.requireSignedData`) is validated against the message already in `StreamMessageValidator`, so that doesn't need to be checked here.
- Unsigned messages can be passed to the `StreamMessageValidator` - in this case the signature will not be checked. Compliance to the chosen policy needs to be checked here though (should throw if `verifySignatures === 'always' and message is not signed)

So the rewritten method could start with the special cases bound to be removed at some point (untested & pseudocode):

```
// Check special cases controlled by the verifySignatures policy
if (options.verifySignatures === 'always' && !msg.signature) {
    throw ...
}
if (options.verifySignatures === 'never' && msg.contentType === StreamMessage.CONTENT_TYPES.MESSAGE) {
    return true
}

// In all other cases validate using the validator
this._verifyStreamMessage(msg)
```

Reply:
Done.